### PR TITLE
[Feature][FlatJson] support flat json type inference core

### PR DESCRIPTION
**Pull request by Seaven**
## What I'm doing:

* move JsonFlater to a class
* support type inference in JsonFlater `deriver_path`
* JsonColumn add flat_type
* ColumnAccessPath add type info

Follow #42787

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by wangsimo0 — DISMISSED** [review]

File: be/src/util/json_flater.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#include "column/vectorized_fwd.h"
+#include "common/status.h"
+#include "types/logical_type.h"
+#include "velocypack/vpack.h"
+
+namespace starrocks {
+namespace vpack = arangodb::velocypack;
+
+class JsonFlater {
```
Comment (kangkaisen) [inline be/src/util/json_flater.h:33]: Flatten

File: be/src/util/json_flater.cpp [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        VLOG(8) << "flat json, internal column too less: " << derived_maps.size()
+                << ", at least: " << config::json_flat_internal_column_min_limit;
+        return;
+    }
+
+    // sort by hit, casts
+    std::vector<pair<std::string_view, FlatColumnDesc>> top_hits(derived_maps.begin(), derived_maps.end());
+    std::sort(top_hits.begin(), top_hits.end(),
+              [](const pair<std::string_view, FlatColumnDesc>& a, const pair<std::string_view, FlatColumnDesc>& b) {
+                  // check hits, the higher the hit rate, the higher the priority.
```
Comment (kangkaisen) [inline be/src/util/json_flater.cpp:279]: what's the hit meaning?

File: be/src/column/column_access_path.cpp [hunk 3]
```diff
@@ -93,6 +98,7 @@ StatusOr<ColumnAccessPathPtr> ColumnAccessPath::convert_by_index(const Field* fi
     path->_from_predicate = this->_from_predicate;
     path->_absolute_path = this->_absolute_path;
     path->_column_index = index;
+    path->_value_type = this->_value_type;
```
[the remark this answers is not in the document]
Comment (Seaven) [reply be/src/column/column_access_path.cpp:101]: `type` is path type(INDEX, KEY, VALUE, FIELD), `value_type` is field data type

File: be/src/util/json_flater.h (hunk 1 above)
Comment (Seaven) [reply be/src/util/json_flater.h:33]: `Flatten` is verb, so use `Flater`

File: be/src/util/json_flater.cpp (hunk 2 above)
Comment (Seaven) [reply be/src/util/json_flater.cpp:279]: we can extract columns with a certain proportion(<= 30%) of null/none values, but columns with a lower proportion of null/none values have higher priority.

File: be/src/util/json_flater.cpp [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+std::vector<LogicalType> JsonFlater::get_flat_types() {
+    std::vector<LogicalType> types;
+    for (const auto& t : _flat_types) {
+        types.emplace_back(JSON_BITS_TO_LOGICAL_TYPE.at(t));
+    }
+    return types;
+}
+
+struct FlatColumnDesc {
+public:
```
Comment (silverbullet233) [inline be/src/util/json_flater.cpp:203]: public keywords can be removed

File: be/src/util/json_flater.cpp [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+    uint8_t type = JsonFlater::JSON_NULL_TYPE_BITS;
+    uint64_t hits = 0;
```
Comment (silverbullet233) [inline be/src/util/json_flater.cpp:205]: better to add some comments for hits and casts.

File: be/src/column/column_access_path.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Seaven) [reply be/src/column/column_access_path.cpp:101]: Add comment

File: be/src/util/json_flater.cpp (hunk 2 above)
Comment (Seaven) [reply be/src/util/json_flater.cpp:282]: add comment

**Review by silverbullet233 — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**